src/types/timestamp.rs

```diff
@@ -865,3 +900,135 @@ mod timestamp_tests {
         assert_eq!(123456789, super::first_n_digits_of(9, 123456789));
     }
 }
+
+#[cfg(test)]
+mod from_ionc_tests {
+    use super::*;
+    use bigdecimal::BigDecimal;
+    use ion_c_sys::timestamp as ionc_ts;
+    use rstest::*;
+
+    fn frac(lit: &str) -> ionc_ts::Mantissa {
+        ionc_ts::Mantissa::Fraction(BigDecimal::parse_bytes(lit.as_bytes(), 10).unwrap())
+    }
+
+    fn decimal(lit: &str) -> Decimal {
+        BigDecimal::parse_bytes(lit.as_bytes(), 10).unwrap().into()
+    }
+
+    /// Creates an `IonDateTime` or panics from constituent components.
+    fn ionc_dt(dt_lit: &str, precision: TSPrecision, offset_kind: TSOffsetKind) -> IonDateTime {
+        let dt = DateTime::parse_from_rfc3339(dt_lit).unwrap();
+        IonDateTime::try_new(dt, precision, offset_kind).unwrap()
+    }
+
+    #[rstest]
+    #[case::year(
+        ionc_dt(
+            "2020-01-01T00:01:00Z",
+            ionc_ts::TSPrecision::Year,
+            ionc_ts::TSOffsetKind::UnknownOffset
+        ),
+        Timestamp::with_year(2020).build().unwrap()
+    )]
+    #[case::month(
+        ionc_dt(
+            "2020-01-01T00:01:00Z",
+            ionc_ts::TSPrecision::Month,
+            ionc_ts::TSOffsetKind::UnknownOffset
+        ),
+        Timestamp::with_year(2020)
+            .with_month(1)
+            .build()
+            .unwrap()
+    )]
+    #[case::day(
+        ionc_dt(
+            "2020-01-01T00:01:00Z",
+            ionc_ts::TSPrecision::Day,
+            ionc_ts::TSOffsetKind::UnknownOffset
+        ),
+        Timestamp::with_year(2020)
+            .with_month(1)
+            .with_day(1)
+            .build()
+            .unwrap()
+    )]
+    #[case::minute_unknown(
+        ionc_dt(
+            "2020-01-01T00:01:00Z",
+            ionc_ts::TSPrecision::Minute,
+            ionc_ts::TSOffsetKind::UnknownOffset
+        ),
+        Timestamp::with_ymd(2020, 1, 1)
+            .with_hour_and_minute(0, 1)
+            .build_at_unknown_offset()
+            .unwrap()
+    )]
+    #[case::minute_minus0800(
+        ionc_dt(
+            "2020-01-01T00:01:00-08:00",
+            ionc_ts::TSPrecision::Minute,
+            ionc_ts::TSOffsetKind::KnownOffset
+        ),
+        Timestamp::with_ymd(2020, 1, 1)
+            .with_hour_and_minute(0, 1)
+            .build_at_offset(8 * 60 * 60)
+            .unwrap()
```

Comment:
This offset does not seem right, and I cut #190 to explain why I think so.  @zslayton might want to look at this to correct me if I misunderstood the API here.